doc/designs/quic-design/connection-id-cache.md

```diff
@@ -0,0 +1,251 @@
+```
+
+### Remote Connection ID APIs
+
+```c
+typedef struct quic_remote_conn_id_cache_st QUIC_REMOTE_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a remote connection ID cache
+ */
+QUIC_REMOTE_CONN_ID_CACHE *ossl_quic_remote_conn_id_cache_new_conn(
+        unsigned int id_limit   /* [active_connection_id_limit] */
+    );
+void ossl_quic_remote_conn_id_cache_free(QUIC_REMOTE_CONN_ID_CACHE *cache);
+
+/*
+ * Add a remote connection ID to the cache
+ */
+int ossl_quic_cache_add_remote_conn_id(QUIC_REMOTE_CONN_ID_CACHE *cache,
+                                       const QUIC_CONNECTION *conn,
+                                       const unsigned char *conn_id,
+                                       size_t conn_id_len,
+                                       uint64_t seq_no);
+
+/*
+ * Query a remote ID for a connection.
+ * If there are multiple non-retired IDs one will be returned in a
+ * non-specified manner.
+ */
+int ossl_quic_conn_id_cache_get0_remote_conn_id(
+        const QUIC_REMOTE_CONN_ID_CACHE *cache,
+        const QUIC_CONNECTION *conn, QUIC_CONN_ID **cid);
+
+/*
+ * Retire remote connection IDs up to and including the one determined by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_remote_conn_ids(
+        QUIC_REMOTE_CONN_ID_CACHE *cache, uint64_t seq_no);
+
+/*
+ * Remove remote connection IDs up to and including the one determined by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_remove_conn_ids(
+        QUIC_REMOTE_CONN_ID_CACHE *cache, uint64_t seq_no);
+```
+
+### Local Connection ID APIs
+
+```c
+typedef struct quic_local_conn_id_cache_st QUIC_LOCAL_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a local connection ID cache
+ */
+QUIC_LOCAL_CONN_ID_CACHE *ossl_quic_local_conn_id_cache_new_conn(void);
+void ossl_quic_local_conn_id_cache_free(QUIC_LOCAL_CONN_ID_CACHE *cache);
+
+/*
+ * Generate a new random local connection ID and associate it with a connection.
+ * For MVP this could just be a zero length ID.
+ */
+int ossl_quic_local_conn_id_cache_new_conn_id(QUIC_LOCAL_CONN_ID_CACHE *cache,
+                                              QUIC_CONNECTION *conn,
+                                              QUIC_CONN_ID **cid);
+
+/*
+ * Remove a local connection and all associated cached IDs
+ */
+int ossl_quic_remove_connection(QUIC_LOCAL_CONN_ID_CACHE *cache,
+                                const QUIC_CONNECTION *conn);
+
+/*
+ * Lookup a local connection by ID.
+ * Returns the connection or NULL if absent.
+ */
+QUIC_CONNECTION *ossl_quic_get0_local_conn(
+        const QUIC_LOCAL_CONN_ID_CACHE *cache,
+        const unsigned char *conn_id, size_t conn_id_len);
+
+/*
+ * Retire local connection IDs up to and including the one specified by the
+ * sequence number.
+ */
+int ossl_quic_conn_id_cache_retire_local_conn_ids(
+        QUIC_LOCAL_CONN_ID_CACHE *cache, uint64_t from_seq_no);
+
+/*
+ * Remove local connection IDs up to and including the one specified by the
```

Comment:
What is the difference between retire and remove?

Reply:
One end sends a retire request to the other.  At this point it cannot send any more packets with that ID but it must still accept packets with its corresponding ID -- it "retires" the connection.  Once the remote end confirms the retirement, no more packets can be sent under these IDs and the cache entry can be removed completely.

Once either end receives a retirement request, it sends confirmation and removes the connection.

This is complicated a bit because both new connection ID and retire connection ID frames can start or finish this process.  The RFC doesn't explain this all that clearly IMO but that's the upshot of 5.1.2.

Reply:
Added a section about ID retirement.